odoo/addons/base/models/ir_profile.py

```diff
@@ -0,0 +1,77 @@
+import json
+import base64
+import datetime
+
+from odoo import fields, models, api
+from odoo.exceptions import UserError
+from odoo.http import request
+from odoo.tools.profiler import make_session
+from odoo.tools.speedscope import Speedscope
+
+class IrProfile(models.Model):
+    _name = 'ir.profile'
+    _description = 'Ir profile'
+    _order = 'session desc, id asc'
+
+    session = fields.Char('Session', index=True)
+    name = fields.Char('Description')
+    duration = fields.Float('Duration')
+
+    init_stack_trace = fields.Text('init_stack_trace', prefetch=False)
+
+    sql = fields.Text('Sql', prefetch=False)
+    traces_async = fields.Text('Traces Async', prefetch=False)
+    traces_sync = fields.Text('Traces Sync', prefetch=False)
+
+    speedscope = fields.Binary('Speedscope', compute='_compute_speedscope')
+    speedscope_url = fields.Text('Open', compute='_compute_url')
+
+    @api.autovacuum
+    def _gc_profile(self):
+        domain = [('create_date', '<', fields.Datetime.now() - datetime.timedelta(days=30))]
+        return self.sudo().search(domain).unlink()
+
+    def _compute_speedscope(self):
+        for execution in self:
+            sp = Speedscope(init_stack_trace=json.loads(execution.init_stack_trace))
+            if execution.sql:
+                sp.add('sql', json.loads(execution.sql))
+            if execution.traces_async:
+                sp.add('frames', json.loads(execution.traces_async))
+            if execution.traces_sync:
+                sp.add('settrace', json.loads(execution.traces_sync))
+
+            execution.speedscope = base64.b64encode(json.dumps(sp.add_default().make()).encode('utf-8'))
+
+    def _compute_url(self):
+        url_root = request.httprequest.url_root
+        for profile in self:
+            content_url = f'{url_root}web/content/{self._name}/{profile.id}/speedscope'
+            profile.speedscope_url = f'/base/static/lib/speedscope/index.html#profileURL={content_url}'
+
+    def _profiling_enabled(self):
+        return self.env['ir.config_parameter'].sudo().get_param('base.profiling_enabled') # change to duration?
+
+    def _toggle_profiling(self, profile=None, config=False):
+        if profile:
+            if self._profiling_enabled():
+                if not request.session.profile_session:
+                    request.session.profile_session = make_session(self.env.user.name)
+            else:
+                request.session.profile_session = False
+                raise UserError('Profiling is not enabled on this database')
```

Comment:
Suggested code structure:
```suggestion
            if not self._profiling_enabled():
                raise UserError('Profiling is not enabled on this database')
            if not request.session.profile_session:
                request.session.profile_session = make_session(self.env.user.name)
```

Reply:
I'm actually not yet checking if the profiling is enabled when profiling, only the state of the session. This means that with your version, if the profiling is disabled in preference but already active on a session, It is impossible to disable profiling on the session using the interface. 

Actually we should check if the profiling is enabled database side at each request if the session has an active profiling, and automaticaly remove it if not. 
(Anyway raising an error if you want to disable the profiling doesn't make sense IMHO)